proposals/devcontainer-collections.md

```diff
@@ -0,0 +1,110 @@
+- Contains metadata for each definition/feature, repository origin, download link, checksum or signing information.
+- All referenced definitions/features would belong to the same user/organization. (e.g. download links would be the same organization in GitHub.)
+- References exact releases of each definition/feature.
+
+| Property | Type | Description |
+| :--- | :--- | :--- |
+| publisher | string | The organization/user that publishes this group of definitions/features. Must match with the repository user in github.|
+| repository | string | The repository where this collection is contained.|
+| version | string | Version information of this collection list.|
+| features | array | The list of features that are contained in this collection. Same information as the metadata file for the feature.|
+| definitions | array | The list of definitions contained in this collection. Same information as the metadata file for the definition.|
+
+Creating the collections file would make some changes over the metadata of the features/definitions where they are generated.
+
+- Ids would get the collection repository added to them. (e.g. for 'https://github.com/community/features' then 'community/features' to 'myfeature1' for a feature with that id)
+- The links to the release files would be added.
+- Date.
+
+The same way creating the definitions/feature file would also be part of the release process for it so that the following metadata can be added:
+- Version.
+- Checksum.
+- Signatures.
+- Date.
+
+## Definitions/features (devcontainer-feature.json/devcontainer-definition.json)
+
+### Common Properties
+
+| Property | Type | Description |
+| :--- | :--- | :--- |
+| id | string | Id of the feature/definition. The ids should be unique in the contest of the repository where they exist. |
+| displayName | string | Name of the feature/definition |
+| shortDescription | string | Optional one sentence description |
+| description | string | Description of the feature/definition |
+| repositoryPath | string | Path to the repository and folder that contains the root of the code for the feature/definition. |
+| keywords | array | List of keywords relevant to a user that would search for this definition/feature. |
+| version | string | Versioning information for the feature/definition. |
+| files | array | List of links for the download files with checksums and optional signature information. |
+
+### Definitions
+
+| Property | Type | Description |
+| :--- | :--- | :--- |
+| type | string | singleContainer/orchestrated |
+| categories | string | The categories where it belongs |
+| architectures | string | Supported architectures |
+| baseOs | string | The operating system in which the definition is based |
+| options | array | List of user selected options at the moment the user selects them in a client application. |
+| styles | array | Variations of the generated definition that contain different information. (e.g. with example code or data.) |
```

Comment:
Isn't this what we call "variants" today? Why styles? Right now this just drives the Dockerfile, bit in concept it could be broader which this seems to allude to.  Or is this intended to be more of a "type"?

Reply:
I think we'll want to stick to the current naming convention for `features`, so an `options` object like this:

```json
         "options": {
            "VARIANT": {
              "type": "string",
              "proposals": [
                "1", 
                "1.16", 
                "1.17", 
                "1-bullseye", 
                "1.16-bullseye", 
                "1.17-bullseye", 
                "1-buster", 
                "1.16-buster", 
                "1.17-buster"
              ],
              "default": "1-bullseye"
            },
            "NODE_VERSION": {
              "type": "string",
              "proposals": [
                "none",
                "lts/*",
                "16",
                "14",
                "12",
                "10"
              ],
              "default": "none"
            }
          }
        },
```

Reply:
Ah - yeah if this is options, We could, however, enable referencing an image directly in this model as well, where an option just updates the "image" field rather than args.  We could have a type of "image" and "buildArg" to distinguish between the two.

Reply:
Changed to variants, I forgot it already existed. The idea is more of a declaration of different release files that correspond to the same definition. Added some more information.